### PR TITLE
Reconnect client if no pong answer

### PR DESCRIPTION
**Pull request by surik**
If no pong answer on ping client connection will be reestablished 

Relate to #62

### CODE REVIEW

File: src/hello_client.erl [hunk 1]
```diff
@@ -165,12 +166,17 @@ handle_call(terminate, _From, State) ->
 %% @hidden
 handle_info({?INCOMING_MSG, Message}, State) ->
     incoming_message(Message, State);
+handle_info(?PING, State = #client_state{ping_counter = PingCounter}) when PingCounter >= 3 ->
+    ?LOG_ERROR("There is no PONG answer on PING for a long time. Client will be restarted.", []),
```
Comment (liveforeverx) [inline src/hello_client.erl]: @ContentSafe Please, check, what additional information(meta fields) or reword needed here.

Comment (thz) [inline src/hello_client.erl]: "for a long time" - seriously? can we please be more specific?
it seems you are just pretending time-measurement but instead you are counting missed ping/pong.
how about being precise and honest in one go? keep track of the timestamp of the last successful ping/pong and compare that to now(). you will have a precise time-period to expose for the log-reader and the tolerance can be configured directly as a time-span (instead of some combination of PingCounter and PingInterval).

File: src/transports/hello_http_client.erl [hunk 2]
```diff
@@ -95,13 +95,13 @@ content_type(Signarute) ->
 http_send(Client, Request, Signarute, State = #http_state{url = URL, options = Options}) ->
     #http_options{method = Method, ib_opts = Opts} = Options,
     {ok, Vsn} = application:get_key(hello, vsn),
-    Headers = [{<<"Content-Type">>, content_type(Signarute)},
-               {<<"Accept">>, content_type(Signarute)},
-               {<<"User-Agent">>, <<"hello/", (list_to_binary(Vsn))/binary>>}],
+    Headers = [{<<"content-type">>, content_type(Signarute)},
+               {<<"accept">>, content_type(Signarute)},
+               {<<"user-agent">>, <<"hello/", (list_to_binary(Vsn))/binary>>}],
```
Comment (thz) [inline src/transports/hello_http_client.erl]: are these changes really in the scope of this PR?
does the change's description and purpose cover these changes?
it seems to me that you are sneaking in unrelated fixes here.

Comment (surik) [inline src/transports/hello_http_client.erl]: @thz there was [a bug with signature for HTTP](https://github.com/travelping/hello/pull/64/files#diff-b07964bc23a58e8323a0399e93364a17R104) and tests hadn't shown it, when I began working on this PR I had noticed this and decided to change other headers to one code style. Sorry for no description, I'm going to add some information to PR message.

File: src/transports/hello_http_client.erl [hunk 3]
```diff
@@ -95,13 +95,13 @@ content_type(Signarute) ->
...
     case hackney:Method(URL, Headers, Request, Opts) of
         {ok, Success, RespHeaders, ClientRef} when Success =:= 200; Success =:= 201; Success =:= 202 ->
             {ok, Body} = hackney:body(ClientRef),
-            Signarute1 = proplists:get_value(<<"Content-Type">>, RespHeaders, <<"undefined">>),
+            Signarute1 = proplists:get_value(<<"content-type">>, RespHeaders, <<"undefined">>),
```
Comment (surik) [inline src/transports/hello_http_client.erl]: Maybe use
`Signature1 = hackney_headers:get_value(<<"content-type">>, hackney_headers:new(RespHeaders), <<"undefined">>).`
for get rid of case sensitive problem

File: src/hello_client.erl (hunk 1 above)
Comment (surik) [inline src/hello_client.erl]: @thz in this log message I wanted to be more clearer: we sent PONG 3 times and didn't receive any answers. For me,  `3 * keep_alive_interval` is a long time for one PONG answer. 
Of course, we could more specific and have timestamp for last PONG message (if there was).
And of course, we could get rid of PingCounter and PIngInterval combination and have something like PingNoAnswerInterval option, which parametrizes time-span in seconds to shutdown if there no PONG answer within. But, in my opinion, PingCounter is more OTP way (like supervisor restart politic)
I maybe don't understand you on 100%, can you give me more examples?

Comment (skattelmann) [inline src/hello_client.erl]: @surik at least the time hello_client was waiting for a $PONG should be logged when the PingCounter is >=3. Assuming the $PING is sent approximately right away from the client the resulting time measurement 3 * keep_alive_interval should be fine. Something like "Ping request to <IP> timed out after 600s. Client will be restarted." is imho ok. Additional meta fields like server IP or the name of the PID (if named) would be very nice but this stuff  is atm also not part of the other logging (see hello_log.hrl). We should fix that in the future.

Comment (thz) [inline src/hello_client.erl]: i do not see any added value by doing 3x ping instead of doing one. a reconnection should be triggered when a single ping/pong gets a delay greater then the threshold.

File: src/transports/hello_http_client.erl (hunk 2 above)
Comment (thz) [inline src/transports/hello_http_client.erl]: i would prefer to split it into two commits and PRs. a PR is the semantic grouping of commits implementing a change.
a PR is not a grouping of unrelated commits.

**Comment by surik** [comment]
I've reworked PR

File: src/hello_client.erl [hunk 4]
```diff
@@ -165,12 +166,16 @@ handle_call(terminate, _From, State) ->
 %% @hidden
 handle_info({?INCOMING_MSG, Message}, State) ->
     incoming_message(Message, State);
+handle_info(?PING, State = #client_state{ping_sent = true, keep_alive_interval = KeepAliveInterval}) ->
+    ?LOG_ERROR("There is no PONG answer on PING for ~p msec. Client will be restarted.", [KeepAliveInterval]),
```
Comment (thz) [inline src/hello_client.erl]: @skattelmann, please!

File: src/hello_client.erl [hunk 5]
```diff
@@ -165,12 +166,16 @@ handle_call(terminate, _From, State) ->
 %% @hidden
 handle_info({?INCOMING_MSG, Message}, State) ->
     incoming_message(Message, State);
+handle_info(?PING, State = #client_state{waiting_for_pong = true, keep_alive_interval = KeepAliveInterval}) ->
+    ?LOG_ERROR("There is no PONG answer on PING for ~p msec. Client will be restarted.", [KeepAliveInterval]),
```
Comment (thz) [inline src/hello_client.erl]: "Client will be restarted"? Isn't it more the *connection* that will be *reestablished*?
*For ~p msec*, seems wrong too, because the argument is only the "KeepAliveInterval" and not elapsed time since last PING (now - last_ping). So when there was no PONG since 7 seconds, but the Interval is 5 seconds, it will incorrectly say "5 seconds", right?

Comment (surik) [inline src/hello_client.erl]: @thz 1) we are restarting client process. if would be more accurate, we are stopping client process and supervisor will restart it if it possible. Maybe we should write "Client will be stopped" and have log message when client starts? 2) Why we should wait for 7 seconds if KeepAliveInterval is 5? When KeepAliveInterval is expired we stop process (actually, we try to send new PING message and realize that there is no PONG and we stop process) see [hello_client.erl](https://github.com/travelping/hello/blob/master/src/hello_client.erl#L172). Of course, these values more approximately but not like 5 and 7 seconds.

Comment (skattelmann) [inline src/hello_client.erl]: There are two things which are kind of irritating here:
1. The client process is restarted and not only the transport handler which is in our case the zmq or http client. The problem which should be fixed here is transport specific.
2.  It would be nice to get to whole time we are waiting for a $PONG since the first $PING where we got no $PONG answer. Currently we only know that a $PING wasn't answered in the KeepAliveInterval. Of course this time can include several restarts of the transport client.

Comment (surik) [inline src/hello_client.erl]: @skattelmann 
1) Transport specific thing will happen in transport specific modules. When we are stopping client with error reason we call transport_terminate/2 and in zmq case it will close socket.
2) Do you mean that we need to save timestamp between client restarts just for logging it?  

Maybe we should think about some metrics for avg/min/max time for ping-pong?

Comment (skattelmann) [inline src/hello_client.erl]: to 1) yes i know, but the whole client state is also dismissed =)
to 2) yes that would be nice. I know this sounds a bit over the top but knowing the time of a connection loss can be crucial in troubleshooting. If you just restart the client transport you could save such a timestamp in the client state and restart the transport without loosing it. 

What do you exactly want to do with these metrics? Just log them? For the timeout we should define a certain 'KeepAliveTimeout' Interval or whatever to be clear about when the client reestablishes the connection.

Comment (liveforeverx) [inline src/hello_client.erl]: @skattelman, I disagree little bit, that it is a job of hello to know, what time the connection was dropped. The connection should write some metric, like alive or not to exometer and some other aggregation system can do graph it and take the time, where it happens first time. There are ready and simple tools to do it, to not make from hello complex, where it possible to reuse existing tools for the same reasons.

Comment (skattelmann) [inline src/hello_client.erl]: "The connection should write some metric, like alive or not"
I thought this is atm not possible with ezmq? If it is possible I'm perfectly fine with your proposal.

Comment (mgumz) [inline src/hello_client.erl]: so, a user of a tiny tiny tiny little hello-service would need the hello-service and then grafana (or the like) to get the very very tiny information for how long the other service did not answer?

really? is it such a hard problem to remember the last time the hello service has received an answer for a pong? we are talking about one 64bit value, at most.

Comment (surik) [inline src/hello_client.erl]: @mgumz this is not about hard or easy to remember value. This is about what we need to do in this unexpected case: crash and restart or just reconnect socket. I see that second way is more popular :)

But I still think metrics for ping/pong are good although it out of this PR.

Comment (mgumz) [inline src/hello_client.erl]: currently we are talking about what to write into the log-message...

**Comment by surik** [comment]
Add time of last received pong to log message and now we are reestablishing connection instead of restart client process
